### ISSUE TITLE
integration with an embeddable go source compiler

### ISSUE THREAD

**Issue by gedw99** [tetratelabs/wazero-playground#1]

**Comment by codefromthecrypt** [comment]
I moved this to here as it allows discussions to use multiple issues. Also, I've noticed other projects cloning the go playground idea and sometimes patterns, and we'd be amiss to not start with something simple like that. As we know, working code is a great way to test out ideas!

https://go.dev/blog/playground
https://github.com/golang/playground
https://github.com/wa-lang/playground

Feel free to start anything here, though be careful about any literal copy/pasting due to license/citation concerns. Also, while this would start with tinygo it is possible there will be other compilers of note, so I would make anything compiler agnostic.

Note that func-e used to use docker to compile wasm with tinygo with wasm. here's the commit that removed it (as it was complicated https://github.com/tetratelabs/func-e/commit/0a220792c8c9f4bfdfd5bf3259c526e2ab6a4e89)

Meanwhile, feel free to fork new issues or discussions and let's try to cite any insight from prior art on each.

**Comment by gedw99** [comment]
I made an Issue about a debugger, so as to not pollute this issue.  Its here: https://github.com/tetratelabs/wazero-playground/issues/2

**Comment by gedw99** [comment]
> So, on 1 or 4) one issue is how to handle non-standard libraries (ex go.mod etc). So, one approach could be to forbid and another might be to supply a git URL instead of a source file.

Good idea !!

> In general, I think an issue like this would have to keep to the minimum requirements, not maximum. I wouldn't expect this issue to get into things like S3 or other cloud storage, rather a higher level abstraction like OCI or http if anything. Similarly, since this is an issue on the wazero repo and we're already pretty far out of scope, things like runtime debugging could be a bridge too far.

HTTP sounds good.
Debugging - i made an issue jst to get it out of here and put in a placeholder for how w could do it later. Yes a bit of boiling the ocean / bridge to far, but lets see.

> Usually issues have a result of a change on the repo in other words, and I'm thinking this is already a bit beyond that. Until we find a better venue, though.. carry on :D

Yep, its all just to get scoping sorting for now

**Comment by james-lawrence** [comment]
I'm primarily interested in a library that I can just point and go. should follow the same standards as the go proper for import resolution.

a bunch of the ideas in this thread are show stoppers for me in that regard.

1. containers - don't want to have to deal with additional infrastructure for a embeddable compiler; this idea is identical to shelling out to another process; only worse as now you have a giant dependency on the containerization platform that is chosen.
2. any RPC related invocations to another process (again this is no longer an embeddable compiler at that point)
3. resolving the source code should assume local only; i.e. no networked calls to retrieve the source. just standard go local module resolution.

basically whatever this compiler is should concern itself with reading local sources files following standard golang import resolution hueristics and compiling the resulting program. usually with a configurable interface for resolving imports which lets users of the library deal with all the messy details for getting those dependencies on disk locally and the security around those issues.

**Comment by james-lawrence** [comment]
ideally (to me) the api would look something like this:

```golang
// setup wasi options etc
func New() Compiler { /* ... */ }

func (t Compiler) Compile(ctx context.Context, r io.Reader, d io.Writer) error { /* ... */ }
func main() {
   var (
      buf = bytes.NewBuffer(nil)
      example = `package main
 func main() {
    print("hello")
}
`
  )
   c := New(ImportResolver(MyResolver()))
   if err = c.Compile(context.Background(), strings.Reader(example), buf); err != nil {
      log.Fatalln(err)
   }
}
```

**Comment by codefromthecrypt** [comment]
I'll give some feedback from the wazero team.

Is a truly embedded go compiler would need full time work for a period of at least several months before even demo state. The type of code would be cognitively loaded and hard to review, and doing that work would be hard to do while also working on other large work. I can say for the wazero team that we can't dedicate time to something like this without sacrificing wazero 1.0 in the process and/or lead to burn out. We still have quite a lot of things to settle which would also benefit this. For example, wasi doesn't work properly, so adding a new compiler to the mix will add even more pressure. That and TinyGo continues to require attention. That said, after 1.0 and its release ecosystem activity (Feb-March) things will be less stressed.

TL;DR; wazero team won't be starting or digging into someone else's embedded go compiler at least until March. However, when things settle down we'd love to be engaged, regardless of if we lead this or not. I'm sure @ShivanshVij would want to as well. Just now.. it isn't possible.

P.S. Something that fakes by calling tinygo (which would fork processes etc.) could be done easier, e.g. a facade like you've shown, and reviewing that would also be easier. If you want to start that here or anywhere, please do and ping here or me directly. I can help, but not lead it.

**Comment by lthibault** [comment]
Count me in as well!

**Comment by gedw99** [comment]
> Would absolutely love to be a part of this.
> 
> A while back our team realized something like this would be amazing for the ecosystem, but we also realized the sheer amount of work that would be required to implement it.
> 
> We ended up building a cloud compiler (using firecracker MicroVMs) that ends up effectively doing this for us using tinygo + some shenanigans for dependency management, but we’d obviously prefer something we could just embed without needing network connectivity.
> 
> As said above, both myself and folks from my team would be interested in being a part of this in the new year.

That’s what I was thinking of doing also as an interim solution ..

**Comment by gedw99** [comment]
@ShivanshVij just saw scale project.

mare you accepting beta users ? Would like to try this out

**Comment by james-lawrence** [comment]
@codefromthecrypt for what its worth I wasn't suggesting wazero do that embedded work I was mostly responding to the idea that any such thing use containers or RPCs. tinygo is fine for the time being at least it can be done locally. I think this is more of a golang proper problem and any solution should be done using that tooling vs reinventing the entire compilation/linking toolchain.

all that really needs to be done is the IR -> WASI work, embedding the golang runtime for GC, implementing the WASI interfaces.

@ShivanshVij https://github.com/golang/go/issues/40053, https://github.com/golang/go/issues/31105, https://github.com/golang/go/issues/38248 are the main issues blocking golang itself from being able to build such a compiler.

**Comment by gedw99** [comment]
Thanks for the links @james-lawrence 

Would be amazing if this stuff gets into the mainline golang compiler.
